In [ ]:
# Import necessary libraries for date and time manipulation, timezone handling, and displaying formatted text in a notebook
from datetime import datetime
from pytz import timezone
from IPython.display import display, Markdown, HTML

# Track the time when the first cell is executed
first_cell_time = datetime.now()

# Set IST timezone (Indian Standard Time)
ist = timezone('Asia/Kolkata')

# Get the current time in IST for the first cell entry
current_time_ist_first = datetime.now(ist).strftime('%Y-%m-%d %H:%M:%S')

# Display the statement "This is the first cell" in green and bold
display(HTML("<p style='color:green; font-weight:bold;'>This is the first cell. Initialization and setup are done below:</p>"))

# Display the current IST time for the first cell in bold (default color)
display(Markdown(f'**First Cell Entry Time: {current_time_ist_first}**'))


This is the first cell. Initialization and setup are done below:


**Date: 2024-06-05**

**Time: 13:09:34 IST+0530**

In [ ]:
# Import the warnings module
import warnings

# Suppress warnings
warnings.filterwarnings("ignore")


1.	import warnings: This line imports the warnings module, which is a built-in module in Python used to handle warning messages.
2.	warnings.filterwarnings("ignore"): This line sets up a filter to ignore all warning messages. When warnings occur during the execution of the script, they would normally be displayed in the console or log. However, with this line, the filterwarnings function is used to suppress all warning messages.
•	The argument "ignore" is passed to the filterwarnings function, indicating that any warnings generated during the execution should be ignored and not displayed.



In [ ]:
# Import the Pandas library and use the alias 'pd'
import pandas as pd

# Import the 'pickle' module
import pickle

# Import necessary libraries for logistic regression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

'''
(C). Data-Visualization, Code
'''

# (C). This section seems to be related to data visualization or code execution.
# However, it's currently commented out. If there is code related to data visualization
# or any other functionality, it should be included here.

# (C). End of Data-Visualization, Code section


'\n(C). Data-Visualization, Code\n'

1.	import pandas as pd: This line imports the Pandas library with the alias 'pd'. Pandas is a popular data manipulation and analysis library in Python.
2.	import pickle: This line imports the pickle module, which is used for serializing and deserializing Python objects. It can be used to save and load models, among other things.
3.	from sklearn.linear_model import LogisticRegression: This line imports the Logistic Regression model from the scikit-learn library. Logistic Regression is a common algorithm used for classification tasks in machine learning.
4.	from sklearn.model_selection import train_test_split: This line imports the train_test_split function from scikit-learn, which is often used to split a dataset into training and testing sets for machine learning model evaluation.



In [ ]:
# Read the CSV file from the given URL and store it in a DataFrame 'df'
# The CSV file contains heart disease data
df = pd.read_csv('https://raw.githubusercontent.com/9394113857/heart-disease-predictor/raghu/heart-data.csv')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 270 entries, 0 to 269
Data columns (total 14 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   age                                   270 non-null    int64  
 1   sex                                   270 non-null    int64  
 2   chest                                 270 non-null    int64  
 3   resting_blood_pressure                270 non-null    int64  
 4   serum_cholestoral                     270 non-null    int64  
 5   fasting_blood_sugar                   270 non-null    int64  
 6   resting_electrocardiographic_results  270 non-null    int64  
 7   maximum_heart_rate_achieved           270 non-null    int64  
 8   exercise_induced_angina               270 non-null    int64  
 9   oldpeak                               270 non-null    float64
 10  slope                                 270 non-null    int64  
 11  number_of_major_ves

In [ ]:
df.shape

(270, 14)

In [ ]:
# Display the first few rows of the DataFrame to get an overview of the data
df.head()

,age,sex,chest,resting_blood_pressure,serum_cholestoral,fasting_blood_sugar,resting_electrocardiographic_results,maximum_heart_rate_achieved,exercise_induced_angina,oldpeak,slope,number_of_major_vessels,thal,class
0,70,1,4,130,322,0,2,109,0,2.4,2,3,3,present
1,67,0,3,115,564,0,2,160,0,1.6,2,0,7,absent
2,57,1,2,124,261,0,0,141,0,0.3,1,0,7,present
3,64,1,4,128,263,0,0,105,1,0.2,2,1,7,absent
4,74,0,2,120,269,0,2,121,1,0.2,1,1,3,absent


In [ ]:
# Rename the column 'class' to 'target' in the DataFrame 'df'
df.rename(columns={"class": "target"}, inplace=True)


1.	df: This is assumed to be the DataFrame in which the column is to be renamed.
2.	rename(columns={"class": "target"}, inplace=True): This is the Pandas DataFrame method used to rename columns. It takes a dictionary as an argument, where keys are the current column names and values are the new column names. In this case, it renames the column "class" to "target".
•	inplace=True: This parameter is set to True to modify the DataFrame in place, meaning the changes are applied directly to the original DataFrame df. If inplace is not set to True, a new DataFrame with the updated column name will be returned, and the original DataFrame df remains unchanged.
So, after executing this line of code, the DataFrame df will have the column name "class" changed to "target".



In [ ]:
# Display the first few rows of the DataFrame to verify the column name change
print("DataFrame after renaming the 'class' column to 'target':")
df.head()


DataFrame after renaming the 'class' column to 'target':


,age,sex,chest,resting_blood_pressure,serum_cholestoral,fasting_blood_sugar,resting_electrocardiographic_results,maximum_heart_rate_achieved,exercise_induced_angina,oldpeak,slope,number_of_major_vessels,thal,target
0,70,1,4,130,322,0,2,109,0,2.4,2,3,3,present
1,67,0,3,115,564,0,2,160,0,1.6,2,0,7,absent
2,57,1,2,124,261,0,0,141,0,0.3,1,0,7,present
3,64,1,4,128,263,0,0,105,1,0.2,2,1,7,absent
4,74,0,2,120,269,0,2,121,1,0.2,1,1,3,absent


In [ ]:
# Display the entire DataFrame to further inspect the changes
print("Entire DataFrame after renaming the 'class' column to 'target':")
print(df)


Entire DataFrame after renaming the 'class' column to 'target':
     age  sex  chest  resting_blood_pressure  serum_cholestoral  \
0     70    1      4                     130                322   
1     67    0      3                     115                564   
2     57    1      2                     124                261   
3     64    1      4                     128                263   
4     74    0      2                     120                269   
..   ...  ...    ...                     ...                ...   
265   52    1      3                     172                199   
266   44    1      2                     120                263   
267   56    0      2                     140                294   
268   57    1      4                     140                192   
269   67    1      4                     160                286   

     fasting_blood_sugar  resting_electrocardiographic_results  \
0                      0                                     2   

In [ ]:
# Replace categorical values in the 'target' column with numerical values in the DataFrame 'df'
df['target'].replace(['absent', 'present'], [0, 1], inplace=True)


Explanation:
1.	df['target']: This selects the 'target' column in the DataFrame df.
2.	.replace(['absent', 'present'], [0, 1], inplace=True): This is using the replace method to replace specific values in the 'target' column.
•	['absent', 'present']: This is a list of values to be replaced in the 'target' column. In this case, it is replacing 'absent' with 0 and 'present' with 1.
•	[0, 1]: This is a list of replacement values corresponding to the values to be replaced. 'absent' is replaced with 0, and 'present' is replaced with 1.
•	inplace=True: This parameter is set to True to modify the DataFrame in place. The changes are applied directly to the original DataFrame df.
So, after executing this line of code, the categorical values 'absent' and 'present' in the 'target' column will be replaced with numerical values 0 and 1, respectively, in the DataFrame df.



In [ ]:
# Create independent variables (features) by dropping the 'target' column
x = df.drop('target', axis=1)

# Create the dependent variable (target)
y = df['target']


Explanation:
1.	x = df.drop('target', axis=1): This line creates the independent variables (x) by dropping the 'target' column from the DataFrame df. The drop method is used to remove a specified column along the specified axis (axis=1 for columns).
•	x is now a DataFrame containing all columns from df except the 'target' column, representing the independent variables or features.
2.	y = df['target']: This line creates the dependent variable (y) by selecting the 'target' column from the DataFrame df.
•	y is now a Series containing the values of the 'target' column, representing the dependent variable or outcome.
These steps are commonly performed in machine learning workflows. The independent variables (x) are used as features for training a machine learning model, and the dependent variable (y) is the target variable that the model aims to predict. This separation allows for the training and evaluation of a predictive model based on the relationship between the features and the target variable.



In [ ]:
x

,age,sex,chest,resting_blood_pressure,serum_cholestoral,fasting_blood_sugar,resting_electrocardiographic_results,maximum_heart_rate_achieved,exercise_induced_angina,oldpeak,slope,number_of_major_vessels,thal
0,70,1,4,130,322,0,2,109,0,2.4,2,3,3
1,67,0,3,115,564,0,2,160,0,1.6,2,0,7
2,57,1,2,124,261,0,0,141,0,0.3,1,0,7
3,64,1,4,128,263,0,0,105,1,0.2,2,1,7
4,74,0,2,120,269,0,2,121,1,0.2,1,1,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...
265,52,1,3,172,199,1,0,162,0,0.5,1,0,7
266,44,1,2,120,263,0,0,173,0,0.0,1,0,7
267,56,0,2,140,294,0,2,153,0,1.3,2,0,3
268,57,1,4,140,192,0,0,148,0,0.4,2,0,6


In [ ]:
y

,target
0,1
1,0
2,1
3,0
4,0
...,...
265,0
266,0
267,0
268,0


In [ ]:
# Split the dataset into training and validation sets for features and target variable
train_x, valid_x, train_y, valid_y = train_test_split(x, y, test_size=0.3, random_state=35)


Explanation:
1.	train_test_split(x, y, test_size=0.3, random_state=35): This function is from scikit-learn's model_selection module, and it is used to split the dataset into training and validation sets.
•	x: The independent variables (features) are passed as the first argument.
•	y: The dependent variable (target) is passed as the second argument.
•	test_size=0.3: This parameter specifies the proportion of the dataset that should be used for the validation set. In this case, 30% of the data will be reserved for validation.
•	random_state=35: This parameter sets the seed for the random number generator. It ensures reproducibility, meaning that if you run the code with the same seed, you will get the same split each time.
2.	train_x, valid_x, train_y, valid_y: The function returns four variables representing the split datasets:
•	train_x: Training set features (independent variables).
•	valid_x: Validation set features (independent variables).
•	train_y: Training set target variable (dependent variable).
•	valid_y: Validation set target variable (dependent variable).
These variables can then be used to train and evaluate machine learning models on the training and validation sets, respectively.



In [ ]:
# Create an instance of the Logistic Regression model
logr = LogisticRegression()


Explanation:
1.	LogisticRegression(): This is the constructor for the Logistic Regression model in scikit-learn. By calling this constructor, you create an instance of the Logistic Regression model, represented by the variable logr.
Now, the variable logr can be used to fit the model on the training data, make predictions, and evaluate its performance. The model can be trained using the training features (train_x) and target variable (train_y) obtained from the earlier split of the dataset. The specific steps for training and using the model would typically involve calling methods such as fit for training and predict for making predictions.



In [ ]:
# Fit (train) the Logistic Regression model on the training data
logr.fit(train_x, train_y)


LogisticRegression()

Explanation:
1.	logr.fit(train_x, train_y): This line calls the fit method on the Logistic Regression model (logr). The fit method is used to train the model on a given dataset.
•	train_x: The training set features (independent variables) are passed as the first argument.
•	train_y: The training set target variable (dependent variable) is passed as the second argument.
After executing this line, the Logistic Regression model is trained on the provided training data, and the learned parameters are stored within the logr instance. The model is now ready to make predictions on new data or be evaluated on a validation set.



In [ ]:
#new_data=np.array(new_data,dtype='int64')
#new_data=new_data.reshape(1,13)
#xnew_data=pd.DataFrame(new_data)

In [ ]:
# Save the trained Logistic Regression model to a file using pickle
pickle.dump(logr, open('heart_model.pkl', 'wb'))


Explanation:
1.	pickle.dump(logr, open('heart_model.pkl', 'wb')): This line uses the pickle.dump function to serialize and save the trained Logistic Regression model (logr) to a file named 'heart_model.pkl'.
•	logr: The trained model instance is passed as the first argument to be serialized.
•	open('heart_model.pkl', 'wb'): The open function is used to open the file 'heart_model.pkl' in binary write ('wb') mode. This allows writing binary data to the file.
The resulting 'heart_model.pkl' file contains the serialized form of the trained Logistic Regression model. This file can be later loaded using pickle.load to retrieve the trained model and make predictions on new data without retraining.



In [ ]:
# Load the trained Logistic Regression model from the saved file
model = pickle.load(open('heart_model.pkl', 'rb'))


Explanation:
1.	pickle.load(open('heart_model.pkl', 'rb')): This line uses the pickle.load function to deserialize and load a previously saved model from the file 'heart_model.pkl'.
•	open('heart_model.pkl', 'rb'): The open function is used to open the file 'heart_model.pkl' in binary read ('rb') mode. This allows reading binary data from the file.
•	The result of pickle.load is a Python object representing the trained Logistic Regression model, and it is assigned to the variable model.
After executing this line, the model variable contains the deserialized Logistic Regression model, and you can use it to make predictions on new data without the need for retraining.



In [ ]:
# Use the loaded Logistic Regression model to make predictions on the validation set
result1 = model.predict(valid_x)


Explanation:
1.	model.predict(valid_x): This line calls the predict method on the Logistic Regression model (model) to make predictions on the features of the validation set (valid_x).
•	valid_x: The validation set features (independent variables) are passed as the argument to the predict method.
2.	result1: The predictions made by the model on the validation set are stored in the variable result1. These predictions are based on the learned parameters of the previously trained Logistic Regression model.
After executing this line, result1 will contain the predicted values for the validation set. These predictions can be compared with the actual values (valid_y) to evaluate the model's performance.



In [ ]:
result1

array([0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0,
       1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0])

In [ ]:
type(result1)

numpy.ndarray

In [ ]:
valid_x

,age,sex,chest,resting_blood_pressure,serum_cholestoral,fasting_blood_sugar,resting_electrocardiographic_results,maximum_heart_rate_achieved,exercise_induced_angina,oldpeak,slope,number_of_major_vessels,thal
77,39,0,3,94,199,0,0,179,0,0.0,1,0,3
17,53,1,4,140,203,1,2,155,1,3.1,3,0,7
133,64,1,4,120,246,0,2,96,1,2.2,3,1,3
23,47,1,4,112,204,0,0,143,0,0.1,1,0,3
150,41,1,3,112,250,0,0,179,0,0.0,1,0,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...
241,39,1,3,140,321,0,2,182,0,0.0,1,0,3
135,46,0,3,142,177,0,2,160,1,1.4,3,0,3
43,46,1,2,101,197,1,0,156,0,0.0,1,0,7
178,67,1,4,125,254,1,0,163,0,0.2,2,2,7


In [ ]:
newdata = valid_x.head(1)

In [ ]:
# Use the loaded Logistic Regression model to make predictions on new data
result2 = model.predict(newdata)


Explanation:
1.	model.predict(newdata): This line calls the predict method on the Logistic Regression model (model) to make predictions on the features of the new data (newdata).
•	newdata: The features of the new data (independent variables) are passed as the argument to the predict method.
2.	result2: The predictions made by the model on the new data are stored in the variable result2. These predictions are based on the learned parameters of the previously trained Logistic Regression model.
After executing this line, result2 will contain the predicted values for the new data. These predictions can be used for further analysis or decision-making based on the model's understanding of patterns learned during training. It's important to ensure that the features in newdata are consistent with the features the model was trained on.



In [ ]:
result2

array([0])

In [ ]:
# Print the predicted results for the validation set
print("result1 is : \n", result1)

# Print the predicted results for the new data
print("\nresult2 is : \n", result2)


result1 is : 
 [0 1 1 0 0 1 1 0 0 1 0 1 0 0 0 1 1 1 1 0 0 0 0 0 0 0 1 1 1 1 0 1 0 0 0 1 1
 0 0 1 0 0 0 0 1 1 0 1 1 0 0 0 1 1 0 1 0 1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 1
 0 0 0 0 0 0 0]

result2 is : 
 [0]


Explanation:
1.	print("result1 is : \n", result1): This line prints the predicted results (result1) for the validation set. The \n in the string adds a newline character for better formatting.
2.	print("\nresult2 is : \n", result2): This line prints the predicted results (result2) for the new data. Again, the \n in the string adds a newline character for better formatting.
After executing these print statements, you will see the predicted results for the validation set (result1) and the new data (result2) in the console output. These predictions can be compared with the actual values to assess the model's performance on the validation set and to understand the model's predictions on the new data.



# Now its predictions time to test the model:-

In [ ]:
# pip install tabulate


In [ ]:
# Import the Pandas library with the alias 'pd'
import pandas as pd

# Import the 'pickle' module
import pickle

# Import the 'tabulate' module for formatting tables
from tabulate import tabulate


Explanation:
1.	import pandas as pd: This line imports the Pandas library, a powerful data manipulation and analysis library in Python. The alias pd is commonly used for brevity in referencing Pandas functions and objects.
2.	import pickle: This line imports the pickle module, which is used for serializing and deserializing Python objects. It is often used to save and load models or other complex Python objects.
3.	from tabulate import tabulate: This line imports the tabulate function from the tabulate module. The tabulate function is useful for formatting and displaying data in a tabular form. It can be helpful for creating well-organized tables when presenting results or data in a readable manner.
These imports suggest that the code might involve working with data using Pandas, saving or loading serialized objects with pickle, and formatting tables using tabulate.



In [ ]:
# Load the dumped model
model = pickle.load(open('heart_model.pkl', 'rb'))


# This default values or example values are for 0, no heart diseased:-

| S.No | Attribute                                          | Value   |
|------|----------------------------------------------------|---------|
| 1    | Age                                                | 45      |
| 2    | Sex                                                | Male    |
| 3    | Chest Pain Type                                    | 2       |
| 4    | Resting Blood Pressure                             | 130     |
| 5    | Serum Cholesterol Level                            | 220     |
| 6    | Fasting Blood Sugar                                | 0       |
| 7    | Resting Electrocardiographic Results               | 1       |
| 8    | Maximum Heart Rate Achieved                        | 160     |
| 9    | Exercise-Induced Angina                            | 0       |
| 10   | ST Depression Induced by Exercise Relative to Rest | 1.5     |
| 11   | Slope of the Peak Exercise ST Segment              | 2       |
| 12   | Number of Major Vessels Colored by Fluoroscopy     | 0       |
| 13   | Thalassemia                                        | 2       |


In [ ]:
# Load the trained Logistic Regression model from the saved file
model = pickle.load(open('heart_model.pkl', 'rb'))

# Set placeholder values
placeholder_age = 45
placeholder_sex = 1  # 0 for female, 1 for male
placeholder_cp = 2
placeholder_trestbps = 130
placeholder_chol = 220
placeholder_fbs = 0
placeholder_restecg = 1
placeholder_thalach = 160
placeholder_exang = 0
placeholder_oldpeak = 1.5
placeholder_slope = 2
placeholder_ca = 0
placeholder_thal = 2

# Take all inputs from the user
age = int(input(f"Enter age (or press Enter to use placeholder {placeholder_age}): ") or placeholder_age)
sex = int(input(f"Enter sex (0 for female, 1 for male, or press Enter to use placeholder {placeholder_sex}): ") or placeholder_sex)
cp = int(input(f"Enter chest pain type (0-3, or press Enter to use placeholder {placeholder_cp}): ") or placeholder_cp)
trestbps = int(input(f"Enter resting blood pressure (or press Enter to use placeholder {placeholder_trestbps}): ") or placeholder_trestbps)
chol = int(input(f"Enter serum cholesterol level (or press Enter to use placeholder {placeholder_chol}): ") or placeholder_chol)
fbs = int(input(f"Enter fasting blood sugar (0 or 1, or press Enter to use placeholder {placeholder_fbs}): ") or placeholder_fbs)
restecg = int(input(f"Enter resting electrocardiographic results (0-2, or press Enter to use placeholder {placeholder_restecg}): ") or placeholder_restecg)
thalach = int(input(f"Enter maximum heart rate achieved (or press Enter to use placeholder {placeholder_thalach}): ") or placeholder_thalach)
exang = int(input(f"Enter exercise-induced angina (0 or 1, or press Enter to use placeholder {placeholder_exang}): ") or placeholder_exang)
oldpeak = float(input(f"Enter ST depression induced by exercise relative to rest (or press Enter to use placeholder {placeholder_oldpeak}): ") or placeholder_oldpeak)
slope = int(input(f"Enter the slope of the peak exercise ST segment (0-2, or press Enter to use placeholder {placeholder_slope}): ") or placeholder_slope)
ca = int(input(f"Enter number of major vessels (0-3, or press Enter to use placeholder {placeholder_ca}): ") or placeholder_ca)
thal = int(input(f"Enter thalassemia (1-3, or press Enter to use placeholder {placeholder_thal}): ") or placeholder_thal)

# Create a dictionary with the user input
user_input = {
    'age': [age],
    'sex': [sex],
    'cp': [cp],
    'trestbps': [trestbps],
    'chol': [chol],
    'fbs': [fbs],
    'restecg': [restecg],
    'thalach': [thalach],
    'exang': [exang],
    'oldpeak': [oldpeak],
    'slope': [slope],
    'ca': [ca],
    'thal': [thal]
}

# Convert the user input to a DataFrame
user_input_df = pd.DataFrame.from_dict(user_input)

# Reshape the input to be a 2D array
reshaped_input = user_input_df.values.reshape(1, -1)

# Make predictions using the loaded model
prediction = model.predict(reshaped_input)

# Create a table with the user information
table = [
    ['S.No', 'Attribute', 'Value'],
    [1, 'Age', age],
    [2, 'Sex', 'Male' if sex == 1 else 'Female'],
    [3, 'Chest Pain Type', cp],
    [4, 'Resting Blood Pressure', trestbps],
    [5, 'Serum Cholesterol Level', chol],
    [6, 'Fasting Blood Sugar', fbs],
    [7, 'Resting Electrocardiographic Results', restecg],
    [8, 'Maximum Heart Rate Achieved', thalach],
    [9, 'Exercise-Induced Angina', exang],
    [10, 'ST Depression Induced by Exercise Relative to Rest', oldpeak],
    [11, 'Slope of the Peak Exercise ST Segment', slope],
    [12, 'Number of Major Vessels Colored by Fluoroscopy', ca],
    [13, 'Thalassemia', thal]
]

# Print the table
print(tabulate(table, headers='firstrow', tablefmt='fancy_grid'))

# Print the predicted result with formatting
print()
if prediction[0] == 0:

    print(f"\033[92m\033[1mThe user, aged {age}, is not likely to have heart disease.\033[0m")
else:
    print(f"\033[91m\033[1mThe user, aged {age}, is likely to have heart disease.\033[0m")


Enter age (or press Enter to use placeholder 45): 
Enter sex (0 for female, 1 for male, or press Enter to use placeholder 1): 
Enter chest pain type (0-3, or press Enter to use placeholder 2): 
Enter resting blood pressure (or press Enter to use placeholder 130): 
Enter serum cholesterol level (or press Enter to use placeholder 220): 
Enter fasting blood sugar (0 or 1, or press Enter to use placeholder 0): 
Enter resting electrocardiographic results (0-2, or press Enter to use placeholder 1): 
Enter maximum heart rate achieved (or press Enter to use placeholder 160): 
Enter exercise-induced angina (0 or 1, or press Enter to use placeholder 0): 
Enter ST depression induced by exercise relative to rest (or press Enter to use placeholder 1.5): 
Enter the slope of the peak exercise ST segment (0-2, or press Enter to use placeholder 2): 
Enter number of major vessels (0-3, or press Enter to use placeholder 0): 
Enter thalassemia (1-3, or press Enter to use placeholder 2): 
╒════════╤═══════

##################

# This default values or example values are for 1, Yes, heart diseased:-

| S.No | Attribute                                          | Value   |
|------|----------------------------------------------------|---------|
| 1    | Age                                                | 60      |
| 2    | Sex                                                | 1       |
| 3    | Chest Pain Type                                    | 3       |
| 4    | Resting Blood Pressure                             | 140     |
| 5    | Serum Cholesterol Level                            | 260     |
| 6    | Fasting Blood Sugar                                | 1       |
| 7    | Resting Electrocardiographic Results               | 2       |
| 8    | Maximum Heart Rate Achieved                        | 120     |
| 9    | Exercise-Induced Angina                            | 1       |
| 10   | ST Depression Induced by Exercise Relative to Rest | 3.0     |
| 11   | Slope of the Peak Exercise ST Segment              | 2       |
| 12   | Number of Major Vessels Colored by Fluoroscopy     | 1       |
| 13   | Thalassemia                                        | 3       |


In [ ]:
# Load the trained Logistic Regression model from the saved file
model = pickle.load(open('heart_model.pkl', 'rb'))

# Placeholder values for someone with heart disease
placeholder_values_heart_disease = {
    'age': 60,
    'sex': 1,  # 0 for female, 1 for male
    'cp': 3,
    'trestbps': 140,
    'chol': 260,
    'fbs': 1,
    'restecg': 2,
    'thalach': 120,
    'exang': 1,
    'oldpeak': 3.0,
    'slope': 2,
    'ca': 1,
    'thal': 3
}

# Take user inputs or use placeholder values for someone with heart disease
user_input = {key: input(f"Enter {key.capitalize()} (or press Enter to use placeholder {value}): ") or value for key, value in placeholder_values_heart_disease.items()}

# Create a DataFrame with the user input
user_input_df = pd.DataFrame([user_input])

# Make predictions using the loaded model
prediction = model.predict(user_input_df.values)

# Create a table with the user information
table_data_heart_disease = [['S.No', 'Attribute', 'Value']]
table_data_heart_disease.extend([(i + 1, key.capitalize(), value) for i, (key, value) in enumerate(user_input.items())])
table_heart_disease = tabulate(table_data_heart_disease, headers='firstrow', tablefmt='fancy_grid')

# Print the table
print(table_heart_disease)

# Print the predicted result with formatting
print()
if prediction[0] == 0:
    print("\033[92m\033[1mThe user is not likely to have heart disease.\033[0m")
else:
    print("\033[91;1mThe user is likely to have heart disease.\033[0m")

Enter Age (or press Enter to use placeholder 60): 
Enter Sex (or press Enter to use placeholder 1): 
Enter Cp (or press Enter to use placeholder 3): 
Enter Trestbps (or press Enter to use placeholder 140): 
Enter Chol (or press Enter to use placeholder 260): 
Enter Fbs (or press Enter to use placeholder 1): 
Enter Restecg (or press Enter to use placeholder 2): 
Enter Thalach (or press Enter to use placeholder 120): 
Enter Exang (or press Enter to use placeholder 1): 
Enter Oldpeak (or press Enter to use placeholder 3.0): 
Enter Slope (or press Enter to use placeholder 2): 
Enter Ca (or press Enter to use placeholder 1): 
Enter Thal (or press Enter to use placeholder 3): 
╒════════╤═════════════╤═════════╕
│   S.No │ Attribute   │   Value │
╞════════╪═════════════╪═════════╡
│      1 │ Age         │      60 │
├────────┼─────────────┼─────────┤
│      2 │ Sex         │       1 │
├────────┼─────────────┼─────────┤
│      3 │ Cp          │       3 │
├────────┼─────────────┼─────────┤
│    

In [ ]:
# Function to print a formatted table
def print_table(header, data):
    # Calculate column widths
    col_widths = [max(len(str(row[i])) for row in data) for i in range(len(header))]

    # Print the header
    print("┌" + "─" * (sum(col_widths) + len(col_widths) * 3 - 1) + "┐")
    print("│", end="")
    for i, col_name in enumerate(header):
        print(f" {col_name.ljust(col_widths[i])} │", end="")
    print()

    # Print separator
    print("├" + "─" * (sum(col_widths) + len(col_widths) * 3 - 1) + "┤")

    # Print data rows
    for row in data:
        print("│", end="")
        for i, value in enumerate(row):
            print(f" {str(value).ljust(col_widths[i])} │", end="")
        print()

    # Print bottom border
    print("└" + "─" * (sum(col_widths) + len(col_widths) * 3 - 1) + "┘")

# Define header and data
header = ["S.No", "Attribute", "Range (No)", "Range (Yes)"]
data = [
    [1, "Age", "45", "60"],
    [2, "Sex", "Male", "Male"],
    [3, "Chest Pain Type", "2", "3"],
    [4, "Resting Blood Pressure", "130", "140"],
    [5, "Serum Cholesterol Level", "220", "260"],
    [6, "Fasting Blood Sugar", "0", "1"],
    [7, "Resting Electrocardiographic Results", "1", "2"],
    [8, "Maximum Heart Rate Achieved", "160", "120"],
    [9, "Exercise-Induced Angina", "0", "1"],
    [10, "ST Depression Induced by Exercise Relative to Rest", "1.5", "3"],
    [11, "Slope of the Peak Exercise ST Segment", "2", "2"],
    [12, "Number of Major Vessels Colored by Fluoroscopy", "0", "1"],
    [13, "Thalassemia", "2", "3"]
]

# Print the table
print_table(header, data)


┌───────────────────────────────────────────────────────────────────────┐
│ S.No │ Attribute                                          │ Range (No) │ Range (Yes) │
├───────────────────────────────────────────────────────────────────────┤
│ 1  │ Age                                                │ 45   │ 60   │
│ 2  │ Sex                                                │ Male │ Male │
│ 3  │ Chest Pain Type                                    │ 2    │ 3    │
│ 4  │ Resting Blood Pressure                             │ 130  │ 140  │
│ 5  │ Serum Cholesterol Level                            │ 220  │ 260  │
│ 6  │ Fasting Blood Sugar                                │ 0    │ 1    │
│ 7  │ Resting Electrocardiographic Results               │ 1    │ 2    │
│ 8  │ Maximum Heart Rate Achieved                        │ 160  │ 120  │
│ 9  │ Exercise-Induced Angina                            │ 0    │ 1    │
│ 10 │ ST Depression Induced by Exercise Relative to Rest │ 1.5  │ 3    │
│ 11 │ Slope of the Pea

**Let's break down each input attribute for both scenarios and explain the differences between the two:**


**1.	Age:**

•	For the "not diseased" scenario: The individual is 45 years old. This age might be considered relatively young and less prone to heart disease compared to older age groups.

•	For the "diseased" scenario: The individual is 60 years old. This age is considered older and is associated with a higher risk of heart disease. Older age is a significant risk factor for heart disease due to factors such as arterial stiffness and accumulated damage over time.

**2.	Sex:**

•	For both scenarios, the sex of the individual is male (1). While sex itself is not a modifiable risk factor, males tend to have a higher risk of heart disease compared to females, particularly at younger ages.

**3.	Chest Pain Type (Cp):**

•	For the "not diseased" scenario: The chest pain type is 2. Chest pain type 2 is typically categorized as "atypical angina" which may indicate a lower likelihood of heart disease.

•	For the "diseased" scenario: The chest pain type is 3. Chest pain type 3 is categorized as "non-anginal pain" which may indicate a higher likelihood of heart disease. Non-anginal pain is less specific and could be related to various conditions including heart issues.

**4.	Resting Blood Pressure (Trestbps):**

•	For both scenarios, the resting blood pressure is provided in mm Hg. In the "not diseased" scenario, the resting blood pressure is 130, which is within the normal range. In the "diseased" scenario, the resting blood pressure is 140, which is slightly elevated and may indicate hypertension, a risk factor for heart disease.

**5.	Serum Cholesterol Level (Chol):**

•	For both scenarios, the serum cholesterol level is provided in mg/dL. In the "not diseased" scenario, the cholesterol level is 220, which may be considered borderline high. In the "diseased" scenario, the cholesterol level is 260, which is higher and indicates an increased risk of heart disease.

**6.	Fasting Blood Sugar (Fbs):**

•	For both scenarios, the fasting blood sugar level is provided. In the "not diseased" scenario, the blood sugar level is 0, indicating a normal fasting blood sugar level. In the "diseased" scenario, the blood sugar level is 1, which suggests elevated blood sugar levels, potentially indicating diabetes, a significant risk factor for heart disease.

**7.	Resting Electrocardiographic Results (Restecg):**

•	For both scenarios, the resting electrocardiographic results are provided. In the "not diseased" scenario, the result is 1, which may indicate some abnormalities in the heart's electrical activity but is not definitive for heart disease. In the "diseased" scenario, the result is also 2, which may indicate more significant abnormalities and an increased risk of heart disease.

**8.	Maximum Heart Rate Achieved (Thalach):**

•	For both scenarios, the maximum heart rate achieved during exercise is provided. In the "not diseased" scenario, the maximum heart rate achieved is 160, which is relatively high and indicative of good cardiovascular fitness. In the "diseased" scenario, the maximum heart rate achieved is 120, which is lower and may indicate impaired cardiovascular function.

**9.	Exercise-Induced Angina (Exang):**

•	For both scenarios, the presence of exercise-induced angina is indicated. In the "not diseased" scenario, the individual did not experience exercise-induced angina (0), suggesting better coronary artery function. In the "diseased" scenario, the individual did experience exercise-induced angina (1), indicating potential coronary artery disease.

**10.	ST Depression Induced by Exercise Relative to Rest (Oldpeak):**

•	For both scenarios, the ST depression induced by exercise relative to rest is provided. In the "not diseased" scenario, the value is 1.5, which may indicate some ischemia during exercise but not necessarily indicative of heart disease. In the "diseased" scenario, the value is 3.0, which is higher and suggests more significant ischemia and a higher likelihood of heart disease.

**11.	Slope of the Peak Exercise ST Segment (Slope):**

•	For both scenarios, the slope of the peak exercise ST segment is provided. In the "not diseased" scenario, the slope is 2, which may indicate a gradual or flat slope, suggesting a lower likelihood of heart disease. In the "diseased" scenario, the slope is also 2, indicating a similar pattern.

**12.	Number of Major Vessels Colored by Fluoroscopy (Ca):**

•	For both scenarios, the number of major vessels colored by fluoroscopy is provided. In both scenarios, the value is 0, indicating no major vessels colored, which may suggest better coronary artery health.

**13.	Thalassemia (Thal):**

For both scenarios, the thalassemia category is provided. In the "not diseased" scenario, the thalassemia category is 2, which may indicate a less severe form of thalassemia. In the "diseased" scenario, the thalassemia category is 3, which may indicate a more severe form of thalassemia, potentially contributing to heart disease risk.

**Overview:**

These differences in input values between the two scenarios illustrate various risk factors and indicators that contribute to the likelihood of heart disease. The combination of these factors helps in predicting the probability of heart disease in individuals.


In [ ]:
# https://chat.openai.com/c/ec1b7d1c-0814-403a-a140-6827f14dc117

In [ ]:
# Track the time when the last cell is executed
last_cell_time = datetime.now()

# Get current time in IST for the last cell entry
now_ist_last = datetime.now(ist)

# Format the current date and time for the last cell
formatted_date_last = now_ist_last.strftime('%Y-%m-%d')
formatted_time_last = now_ist_last.strftime('%H:%M:%S %Z%z')

# Display the statement "All cells above have been executed..." in red and bold
display(HTML("<p style='color:red; font-weight:bold;'>All cells above have been executed. Below is the current date and time in IST:</p>"))

# Display the last cell entry time in bold (default color)
display(Markdown(f'**Last Cell Entry Time: {formatted_date_last} {formatted_time_last}**'))


All cells above have been executed. Below is the current date and time in IST:


**Date: 2024-08-27**

**Time: 18:02:41 IST+0530**

In [ ]:
# Calculate the time difference between first and last cell execution
time_diff = last_cell_time - first_cell_time

# Extract hours, minutes, and seconds from the time difference
hours = time_diff.seconds // 3600  # Total seconds divided by 3600 to get hours
minutes = (time_diff.seconds % 3600) // 60  # Remainder divided by 60 to get minutes
seconds = time_diff.seconds % 60  # Remainder gives the number of seconds

# Display the statement for total time taken in execution in orange and bold
display(HTML("<p style='color:orange; font-weight:bold;'>Total Time Taken for Notebook Execution:</p>"))

# Display the first and last cell entry times along with the total time taken in HH:MM:SS format
display(Markdown(f'**First Cell Entry Time: {current_time_ist_first}**'))
display(Markdown(f'**Last Cell Entry Time: {formatted_date_last} {formatted_time_last}**'))
display(Markdown(f'**Time Taken: {hours} hour(s), {minutes} minute(s), and {seconds} second(s)**'))


# Questions and Answers from the Notebook

Here are some questions and answers based on the content of the notebook to help you understand the process and prepare for an interview.

## Q1: What is the purpose of the first few code cells in this notebook?

**A1:** The first few cells are for initial setup and tracking the execution time of the notebook.
1. They import necessary libraries for date and time manipulation (`datetime`, `pytz`).
2. They set the timezone to IST (Indian Standard Time).
3. They record the start time of the notebook execution.
4. They include a cell to suppress warnings.

## Q2: What libraries are imported for data handling and model building?

**A2:** The key libraries imported are:
1. `pandas` (as `pd`) for data manipulation and analysis.
2. `pickle` for saving and loading Python objects, specifically the trained model.
3. `sklearn.linear_model.LogisticRegression` for implementing the logistic regression model.
4. `sklearn.model_selection.train_test_split` for splitting the data into training and testing sets.
5. `tabulate` for formatting the input data into a table.

## Q3: Where does the heart disease data come from and what is its initial format?

**A3:**
1. The data is loaded from a CSV file hosted on GitHub: `https://raw.githubusercontent.com/9394113857/heart-disease-predictor/raghu/heart-data.csv`.
2. The initial format of the data is shown using `df.info()` and `df.head()`.
3. It contains 14 columns with a mix of integer, float, and object data types.
4. There are 270 entries (rows) in the dataset.

## Q4: How is the 'class' column handled in the data preprocessing steps?

**A4:**
1. The 'class' column, which contains categorical values ('absent' and 'present'), is renamed to 'target' using `df.rename()`.
2. The categorical values in the 'target' column are then replaced with numerical values: 'absent' becomes 0 and 'present' becomes 1, using `df['target'].replace()`.

## Q5: How is the data split for training and validation?

**A5:**
1. The data is split into independent variables (features) `x` and the dependent variable (target) `y`. `x` contains all columns except 'target', and `y` contains only the 'target' column.
2. The `train_test_split` function from `sklearn.model_selection` is used to divide the data.
3. The `test_size=0.3` parameter indicates that 30% of the data is allocated for the validation set.
4. The `random_state=35` parameter ensures that the data split is the same every time the code is run, making the results reproducible.

## Q6: Which machine learning model is used and how is it trained and saved?

**A6:**
1. A `LogisticRegression` model from `sklearn.linear_model` is used.
2. The model is trained (fitted) on the training data (`train_x`, `train_y`) using the `logr.fit()` method.
3. The trained model is saved to a file named `heart_model.pkl` using the `pickle.dump()` function in binary write mode (`'wb'`).

## Q7: How is the saved model loaded and used for predictions?

**A7:**
1. The trained model is loaded from the `heart_model.pkl` file using the `pickle.load()` function in binary read mode (`'rb'`).
2. The loaded model is used to make predictions on new data (e.g., `valid_x` and `newdata`) using the `model.predict()` method.

## Q8: How is the user input handled for making predictions on a single instance?

**A8:**
1. The code prompts the user to enter values for each of the 13 features.
2. Placeholder values are provided for each input, and if the user just presses Enter, the placeholder value is used.
3. The user input is collected into a Python dictionary.
4. This dictionary is converted into a pandas DataFrame.
5. The DataFrame values are reshaped into a 2D array (`1, -1`) as required by the `predict` method.
6. The reshaped input is passed to the loaded model's `predict()` method to get the prediction.
7. The input values are presented in a formatted table using the `tabulate` library.
8. The prediction result (0 for no heart disease, 1 for heart disease) is printed with colored and bold text.

## Q9: What is the difference between the two prediction examples provided in the notebook?

**A9:**
1. The notebook shows two examples of making predictions after training and saving the model.
2. The first example (using `valid_x.head(1)`) demonstrates making a prediction on the first row of the validation set, which happens to be predicted as 'absent' (0), indicating no heart disease.
3. The second example (using user input with placeholder values) demonstrates making a prediction for a hypothetical individual with characteristics that are more indicative of heart disease (e.g., older age, higher cholesterol, higher fasting blood sugar, exercise-induced angina), and the model predicts 'present' (1), indicating likely heart disease.
4. The detailed explanation in the markdown cell `5lh4q-jiiLRZ` further highlights the differences in the input attributes between the two scenarios and how they relate to the likelihood of heart disease.

## Q10: How is the total execution time of the notebook calculated and displayed?

**A10:**
1. The time when the first cell starts executing is recorded in the `first_cell_time` variable.
2. The time when the last cell finishes executing is recorded in the `last_cell_time` variable.
3. The difference between `last_cell_time` and `first_cell_time` is calculated to get the `time_diff`.
4. The `time_diff` is then broken down into hours, minutes, and seconds.
5. The total time taken is displayed in a formatted output, including the first and last cell entry times in IST.

# Interview Preparation

Based on this notebook, here are some potential interview questions and how you might answer them:

## General Questions:

1. **Can you walk me through the steps you took in this notebook to build a heart disease predictor?**
   * **Answer:** Explain the process from data loading and preprocessing (renaming 'class' to 'target', converting categorical labels to numerical) to splitting the data, training a Logistic Regression model, saving the model using `pickle`, and then loading it to make predictions on new data.

2. **Why did you choose Logistic Regression for this task?**
   * **Answer:** Mention that Logistic Regression is a simple yet effective algorithm for binary classification problems like predicting the presence or absence of heart disease. You could also mention that it's a good baseline model and the coefficients can be interpreted to understand the influence of different features.

3. **How did you handle the categorical 'class' variable?**
   * **Answer:** Explain the process of renaming the column and then using the `.replace()` method to convert 'absent' and 'present' to 0 and 1 respectively. Emphasize the importance of converting categorical data to numerical format for most machine learning algorithms.

4. **What is the purpose of splitting the data into training and validation sets?**
   * **Answer:** Explain that splitting the data is crucial to evaluate the model's performance on unseen data and prevent overfitting. The model is trained on the training set and evaluated on the validation set.

5. **Why did you use `pickle` to save the model?**
   * **Answer:** Explain that `pickle` is used to serialize the trained model object, allowing you to save its state to a file. This is useful for deploying the model later without having to retrain it.

6. **How would you evaluate the performance of this model?**
   * **Answer:** Discuss common classification evaluation metrics such as accuracy, precision, recall, F1-score, and AUC-ROC. Explain what each metric means and why they are important for evaluating a medical prediction model.

7. **What are some potential limitations of this model or dataset?**
   * **Answer:** Discuss the size of the dataset (270 entries is relatively small). Mention that Logistic Regression is a linear model and might not capture complex non-linear relationships in the data. Discuss the importance of feature scaling (which was not done in this notebook) for some algorithms, although Logistic Regression is less sensitive to it.

## Code-Specific Questions:

8. **Explain the `df.rename(columns={"class": "target"}, inplace=True)` line.**
   * **Answer:** Explain that this line renames the column named "class" to "target" in the DataFrame `df`. `inplace=True` modifies the DataFrame directly without creating a new one.

9. **What does `train_test_split(x, y, test_size=0.3, random_state=35)` do?**
   * **Answer:** Explain that this function splits the features `x` and target `y` into training and validation sets. `test_size=0.3` allocates 30% of the data to the validation set, and `random_state=35` ensures the split is reproducible.

10. **How does the user input section work to make a prediction for a new individual?**
    * **Answer:** Explain how the code prompts the user for each feature value, uses placeholder values if the user doesn't provide input, stores the input in a dictionary, converts it to a pandas DataFrame, reshapes it, and then uses `model.predict()` to get the prediction.

## Further Discussion/Improvements:

11. **How could you improve the accuracy of this model?**
    * **Answer:** Suggest trying other classification algorithms (e.g., Support Vector Machines, Random Forests, Gradient Boosting). Discuss the possibility of feature engineering, handling outliers, or collecting more data.

12. **How would you deploy this model for real-world use?**
    * **Answer:** Discuss potential deployment methods such as creating a web application using Flask or Django, or using cloud platforms like Google Cloud Platform, AWS, or Azure to host the model as an API.

Remember to explain these concepts clearly and confidently, relating them back to the code in the notebook. Good luck!

## Top Questions and Answers for ML Interview

Here are some straightforward questions and answers based on this notebook, covering its purpose, predictions, and real-world relevance.

**Q1: What is the main purpose of this notebook?**

*   **A1:** This notebook demonstrates the process of building, training, and using a machine learning model to predict the likelihood of heart disease based on various health attributes.

**Q2: What specific machine learning task is performed?**

*   **A2:** This notebook performs a binary classification task, predicting whether an individual is likely to have heart disease (represented by 1) or not (represented by 0).

**Q3: What type of machine learning model is used?**

*   **A3:** A Logistic Regression model is used for the prediction.

**Q4: What kind of data is used in this notebook?**

*   **A4:** The notebook uses a dataset containing various health attributes of individuals, such as age, sex, chest pain type, blood pressure, cholesterol, etc., and their corresponding heart disease status.

**Q5: How is the model trained and saved?**

*   **A5:** The data is split into training and validation sets. The Logistic Regression model is trained on the training data using the `fit` method. The trained model is then saved using the `pickle` module for later use.

**Q6: How are predictions made using the saved model?**

*   **A6:** The saved model is loaded using `pickle`. New data (individual health attributes) is then fed into the loaded model's `predict` method to get a prediction (0 or 1).

**Q7: What are some sample use cases for this type of heart disease prediction model?**

*   **A7:**
    *   **Clinical Decision Support:** Assisting doctors in assessing a patient's risk of heart disease based on their health profile.
    *   **Preventive Healthcare:** Identifying individuals at high risk who could benefit from lifestyle changes or early interventions.
    *   **Health and Wellness Apps:** Integrating the model into apps to provide users with personalized heart health risk assessments.

**Q8: Can you give examples of real-world services or products that use similar concepts?**

*   **A8:**
    *   **Electronic Health Records (EHR) Systems:** Some EHR systems incorporate risk prediction models to flag patients who may be at increased risk for certain conditions, including heart disease.
    *   **Telemedicine Platforms:** These platforms might use similar models during virtual consultations to help assess patient risk remotely.
    *   **Wearable Health Devices and Apps:** Some advanced health trackers and apps use collected user data (like heart rate variability) in conjunction with predictive models to offer insights into cardiovascular health.
    *   **Medical Research and Drug Development:** Machine learning models are widely used in research to identify patterns in patient data and predict outcomes, which can inform drug development and clinical trial design.

**Q9: What are the key benefits of using a machine learning model for this prediction?**

*   **A9:**
    *   **Efficiency:** Allows for quick risk assessment based on input data.
    *   **Data-Driven Insights:** Can identify complex relationships between multiple health factors and heart disease risk.
    *   **Scalability:** Once trained, the model can be used to make predictions for a large number of individuals.

**Q10: What are potential future improvements for this notebook or model?**

*   **A10:**
    *   Exploring other machine learning algorithms (e.g., Random Forest, Gradient Boosting, SVM) to potentially improve accuracy.
    *   Performing more extensive data preprocessing, such as feature scaling or handling outliers.
    *   Evaluating the model using additional metrics like precision, recall, F1-score, and AUC-ROC, which are important for medical prediction tasks.
    *   Implementing cross-validation for more robust model evaluation.
    *   Considering techniques for interpreting the model's predictions to understand which features are most influential.